posthog/hogql_queries/insights/retention_appearance_query_runner.py

```diff
@@ -0,0 +1,140 @@
+from datetime import timedelta
+from typing import Any, Dict, List, Optional, Union, cast
+
+from posthog.hogql import ast
+from posthog.hogql.constants import LimitContext, get_max_limit_for_context, get_default_limit_for_context
+from posthog.hogql.parser import parse_select
+from posthog.hogql.timings import HogQLTimings
+from posthog.hogql_queries.insights.paginators import HogQlHasMorePaginator
+from posthog.hogql_queries.insights.retention_query_runner import RetentionQueryRunner, DEFAULT_TOTAL_INTERVALS
+from posthog.hogql_queries.query_runner import QueryRunner, get_query_runner
+from posthog.models import Team
+from posthog.queries.actor_base_query import get_groups, get_people, SerializedGroup, SerializedPerson
+from posthog.schema import HogQLQueryModifiers, RetentionAppearanceQuery, PersonsQueryResponse
+
+
+class RetentionAppearanceQueryRunner(QueryRunner):
+    query: RetentionAppearanceQuery
+    query_type = RetentionAppearanceQuery
+
+    def __init__(
+        self,
+        query: RetentionAppearanceQuery | Dict[str, Any],
+        team: Team,
+        timings: Optional[HogQLTimings] = None,
+        modifiers: Optional[HogQLQueryModifiers] = None,
+        limit_context: Optional[LimitContext] = None,
+    ):
+        super().__init__(query, team=team, timings=timings, modifiers=modifiers, limit_context=limit_context)
+        self.paginator = HogQlHasMorePaginator(limit=self.query_limit(), offset=self.query.offset or 0)
+
+    def query_limit(self) -> int:
+        # TODO: Reuse from somewhere else
+        max_rows = get_max_limit_for_context(self.limit_context)
+        default_rows = get_default_limit_for_context(self.limit_context)
+        return min(max_rows, default_rows if self.query.limit is None else self.query.limit)
+
+    def get_source_query_runner(self) -> RetentionQueryRunner:
+        return cast(RetentionQueryRunner, get_query_runner(self.query.source, self.team, self.timings))
+
+    def to_query(self) -> ast.SelectQuery:
+        placeholders = {
+            "actor_query": self.get_source_query_runner().actor_query(
+                breakdown_values_filter=[self.query.selectedInterval]
+            ),
+        }
+        with self.timings.measure("retention_query"):
+            retention_query = parse_select(
+                """
+                    SELECT
+                        actor_id,
+                        groupArray(actor_activity.intervals_from_base) AS appearances
+
+                    FROM {actor_query} AS actor_activity
+
+                    GROUP BY actor_id
+
+                    -- make sure we have stable ordering/pagination
+                    -- NOTE: relies on ids being monotonic
+                    ORDER BY
+                        length(appearances) DESC,
+                        actor_id
+                """,
+                placeholders,
+                timings=self.timings,
+            )
+        return retention_query
```

Comment:
This could likely go into the `to_actors_query` method inside `RetentionQueryRunner`